pennylane/optimize/lie_gradient.py

```diff
@@ -0,0 +1,352 @@
+
+        The Lie gradient is given by.
+
+        .. math::
+
+             \text{grad}f(U^{(t)}}) = -[U \rho U^\dag, H]
+
+        Subsequent steps of this optimizer will append a Trotterized version of the exact Lie
+        gradient and grow the circuit.
+
+        Args:
+            circuit (qml.QNode): the user-defined hyperparameter :math:`\eta`
+            hamiltonian (qml.Hamiltonian): the user-defined hyperparameter :math:`\eta`
+            stepsize (float): the user-defined hyperparameter :math:`\eta`
+            **kwargs
+        **Examples:**
+
+        >>> @qml.qnode(qml.device("default.qubit", wires=2))
+        ... def quant_fun():
+        ...     qml.RX(0.1, wires=[0])
+        ...     qml.RY(0.5, wires=[1])
+        ...     qml.CNOT(wires=[0,1])
+        ...     qml.RY(0.6, wires=[0])
+        ...     return qml.state()
+        Define a Hamiltonian cost function to minimize.
+        >>> hamiltonian = qml.Hamiltonian(coeffs=[-1.]*3, observables=[qml.PauliX(0), qml.PauliZ(1), qml.PauliY(0)@qml.PauliX(1)])
+        Instatiate the optimizer with the initial circuit and the cost function. Set the stepsize
+        accordingly.
+        >>> opt = LieGradientOptimizer(circuit = quant_fun, hamiltonian=hamiltonian, stepsize=0.1)
+        Applying 10 steps gets us close the ground state of E=-2.23
+        >>> for step in range(10):
+        ...    print(step)
+        ...    state = opt.step_and_cost()
+        ...    print(np.trace(H @ np.outer(state, state.conj())))
+
+        """
+        if not isinstance(circuit, qml.QNode):
+            raise TypeError(
+                f"`circuit` must be a `qml.QNode`, " f"received {type(circuit)} "
+            )
+
+        self.circuit = circuit
+        self.circuit.construct([], {})
+        self.nqubits = max(circuit.device.wires) + 1
+
+        if self.nqubits > 4:
+            print(
+                "WARNING: The exact Lie gradient is exponentially expensive in the number of qubits,"
+                f"optimizing a {self.nqubits} qubit circuit may be slow."
+            )
+
+        (
+            self.lie_algebra_basis_ops,
+            self.lie_algebra_basis_names,
+        ) = self.get_su_n_operators()
+
+        if not isinstance(hamiltonian, qml.Hamiltonian):
+            raise TypeError(
+                f"`hamiltonian` must be a `qml.Hamiltonian`, "
+                f"received {type(hamiltonian)} "
+            )
+
+        self.hamiltonian = hamiltonian
+        self.coeffs, self.observables = self.hamiltonian.terms
+        pl_pauli_to_idx = {"PauliX": 1, "PauliY": 2, "PauliZ": 3}
+        hamiltonian_basis_names = []
+        for op in hamiltonian.ops:
+            name = [0] * self.nqubits
+            if len(op.wires) > 1:
+                for loc, p in zip(op.wires, op.name):
+                    name[loc] = pl_pauli_to_idx[p]
+            else:
+                name[op.wires[0]] = pl_pauli_to_idx[op.name]
+            hamiltonian_basis_names.append(tuple(name))
```

Comment:
I think (correct me if wrong!) that this is equivalent to

```python
pauli_strings = [
    tuple(
        map(
            self.pauli_str_to_int.get,
            qml.grouping.pauli_word_to_string(o, wire_map=dev.wires),
        )
    )
    for o in H.ops
]
```

which might be nice, since it uses existing, tested, functionality

Reply:
Turns out I didnt even need this code :facepalm: